### ISSUE TITLE
memleak.py doesn't record mmap and munmap

### ISSUE THREAD

**Issue by yuzhichang** [iovisor/bcc#1416]
memleak.py currently doesn't record allocations via mmap(). Glibc does this for larger allocations.

**Comment by yonghong-song** [comment]
Agreed. Some allocators like jemalloc heavily use mmap for allocations.

**Comment by markdrayton** [comment]
I'm not sure I understand this — `memleak.py` traces calls to `malloc`, not whatever the allocator is calling behind the scenes. It shouldn't matter if the allocator uses `brk`/`sbrk` or `mmap`.

**Comment by yuzhichang** [comment]
Some C/C++ applications call mmap directly instead of malloc.

**Comment by goldshtn** [comment]
That would be a slightly different scenario then. You could also say that some applications open files, and then forget to close them. I don’t know if all these scenarios fit into memleak - I’m open to considering it, but just pointing out it’s not really memory allocation tracing anymore, but a more general leak tracer.

**Comment by goldshtn** [comment]
@yuzhichang Would you like to try proposing a PR that adds more generic resource tracing to memleak? Or perhaps a separate tool for this purpose?

**Comment by junwang123** [comment]
@goldshtn There is also anonymous mmap with the MAP_ANONYMOUS flag in which no actual file is involved. As the mmap memory is touched, RSS keeps growing for that segment. I'm not sure how to trace that kind of leak though.

**Comment by yuzhichang** [comment]
[`apr_pools.c`](https://github.com/apache/apr/blob/7a52a7c060ebb42f6e83505755b49bc1e1acd183/memory/unix/apr_pools.c) uses mmap(with the MAP_ANONYMOUS flag) or malloc to allocate memory for pool nodes.

```
static APR_INLINE
apr_memnode_t *allocator_alloc(apr_allocator_t *allocator, apr_size_t in_size)
......
    /* If we haven't got a suitable node, malloc a new one
     * and initialize it.
     */
#if APR_ALLOCATOR_GUARD_PAGES
    if ((node = mmap(NULL, size + 2 * GUARDPAGE_SIZE, PROT_NONE,
                     MAP_PRIVATE|MAP_ANON, -1, 0)) == MAP_FAILED)
#elif APR_ALLOCATOR_USES_MMAP
    if ((node = mmap(NULL, size, PROT_READ|PROT_WRITE,
                     MAP_PRIVATE|MAP_ANON, -1, 0)) == MAP_FAILED)
#else
    if ((node = malloc(size)) == NULL)
#endif
        return NULL;
......

```
